androidHyperskillApp/src/main/java/org/hyperskill/app/android/challenge/delegate/ChallengeCardDelegate.kt

```diff
@@ -0,0 +1,66 @@
+package org.hyperskill.app.android.challenge.delegate
+
+import android.content.Context
+import android.widget.Toast
+import androidx.compose.runtime.getValue
+import androidx.compose.ui.platform.ComposeView
+import androidx.fragment.app.FragmentManager
+import androidx.lifecycle.compose.collectAsStateWithLifecycle
+import kotlinx.coroutines.flow.MutableStateFlow
+import org.hyperskill.app.R
+import org.hyperskill.app.android.challenge.ui.ChallengeCard
+import org.hyperskill.app.android.core.extensions.openUrl
+import org.hyperskill.app.android.core.view.ui.dialog.CreateMagicLinkLoadingProgressDialogFragment
+import org.hyperskill.app.android.core.view.ui.dialog.dismissDialogFragmentIfExists
+import org.hyperskill.app.android.core.view.ui.widget.compose.HyperskillTheme
+import org.hyperskill.app.challenges.widget.presentation.ChallengeWidgetFeature
+import org.hyperskill.app.challenges.widget.view.model.ChallengeWidgetViewState
+import org.hyperskill.app.challenges.widget.view.model.isLoadingMagicLink
+import ru.nobird.android.view.base.ui.extension.showIfNotExists
+
+class ChallengeCardDelegate {
+    private val stateFlow: MutableStateFlow<ChallengeWidgetViewState?> = MutableStateFlow(null)
+
+    fun setup(
+        composeView: ComposeView,
+        onNewMessage: (ChallengeWidgetFeature.Message) -> Unit
+    ) {
+        composeView.setContent {
+            HyperskillTheme {
+                val viewState by stateFlow.collectAsStateWithLifecycle()
+                viewState?.let { actualViewState ->
+                    ChallengeCard(viewState = actualViewState, onNewMessage = onNewMessage)
+                }
+            }
+        }
+    }
+
+    fun render(
+        fragmentManager: FragmentManager,
+        state: ChallengeWidgetViewState
+    ) {
+        stateFlow.value = state
+        if (state is ChallengeWidgetViewState.Content && state.isLoadingMagicLink) {
+            CreateMagicLinkLoadingProgressDialogFragment.newInstance()
+                .showIfNotExists(fragmentManager, CreateMagicLinkLoadingProgressDialogFragment.TAG)
+        } else {
+            fragmentManager.dismissDialogFragmentIfExists(CreateMagicLinkLoadingProgressDialogFragment.TAG)
+        }
+    }
+
+    fun handleAction(
+        context: Context,
+        action: ChallengeWidgetFeature.Action.ViewAction
+    ) {
+        when (action) {
+            is ChallengeWidgetFeature.Action.ViewAction.OpenUrl -> {
```

Comment:
`ChallengeWidgetFeature.Action.ViewAction.OpenUrl` has a boolean flag `shouldOpenInApp`:

1. If `true` we should open a URL inside the app. When user clicks on the link in the description to read the challenge rules.
2. If `false` we should open a URL outside the app (Chrome, ...). When user clicks "Collect Reward" button.